File: 算法/C/第3章 排序算法/3.5 归并排序/merge_sort.c
#include <stdio.h>

/**
 * @brief  打印数组
 * @param  arr: 数组
 * @param  n: 数组长度
 * @retval None
 */
void showArr(int *arr, int n) {
    for(int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

/**
 * @brief  合并
 * @param  arr: 待排序数组
 * @param  start: 开始下标
 * @param  mid: 中间下标
 * @param  end: 结束下标
 * @param  temp: 临时数组
 * @retval None
 */
void merge(int *arr, int start, int mid, int end, int *temp) {
    int i = start;
    int j = mid + 1;
    int k = 0;

    while(i <= mid && j <= end) {
        if(arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }

    while(i <= mid) {
        temp[k++] = arr[i++];
    }
    while(j <= end) {
        temp[k++] = arr[j++];
    }

    for(int i = 0; i < k; i++) {
        arr[start+i] = temp[i];
    }
}

/**
 * @brief  归并排序
 * @param  arr: 待排序数组
 * @param  start: 开始下标
 * @param  end: 结束下标
 * @param  temp: 临时数组
 * @retval None
 */
void mergeSort(int *arr, int start, int end, int *temp) {
    if(start < end) {
        int mid = start + (end - start) / 2;
        mergeSort(arr, start, mid, temp);
        mergeSort(arr, mid+1, end, temp);
        merge(arr, start, mid, end, temp);
    }
}

int main() {
    int arr[] = {5, 8, 6, 3, 9, 2, 1, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    int temp[n];        // 临时数组
    
    printf("排序前：");
    showArr(arr, n);
    mergeSort(arr, 0, n-1, temp);
    printf("排序后：");
    showArr(arr, n);
    return 0;
}

File: code/Python/Chapter8/8.6 归并排序/merge_sort.py
def merge(list, start, mid, end, temp):
    """
        合并
        Args:
            list (list): 待排序列表
            start (int): 开始下标
            mid (int): 中间下标
            end (int): 结束下标
            temp (list): 临时列表
    """
    i = start
    j = mid + 1
    k = 0

    while i <= mid and j <= end:
        if list[i] <= list[j]:
            temp[k] = list[i]
            k += 1
            i += 1
        else:
            temp[k] = list[j]
            k += 1
            j += 1

    while i <= mid:
        temp[k] = list[i]
        k += 1
        i += 1
    while j <= end:
        temp[k] = list[j]
        k += 1
        j += 1

    for i in range(k):
        list[start+i] = temp[i]

def merge_sort(list, start, end, temp):
    """
        归并排序
        Args:
            list (list): 待排序列表
            start (int): 开始下标
            end (int): 结束下标
            temp (list): 临时列表
    """
    if start < end:
        mid = start + (end - start) // 2
        merge_sort(list, start, mid, temp)
        merge_sort(list, mid+1, end, temp)
        merge(list, start, mid, end, temp)

def main():
    list = [5, 8, 6, 3, 9, 2, 1, 7]
    temp = [0] * len(list)       # 临时列表
    print("排序前：" + str(list))
    merge_sort(list, 0, len(list)-1, temp)
    print("排序后：" + str(list))

if __name__ == "__main__":
    main()